SetupDataPkg/Library/ConfigVariableListLib/ConfigVariableListLib.c

```diff
@@ -15,9 +15,228 @@
+  CONST CONFIG_VAR_LIST_HDR  *VarList = NULL;
+  UINTN                      BinSize  = 0;
+  EFI_STATUS                 Status   = EFI_SUCCESS;
+  CHAR16                     *VarName = NULL;
+  CHAR8                      *Data    = NULL;
+  UINT32                     Attributes;
+  UINT32                     CRC32;
+  UINT32                     CalcCRC32;
+  UINT32                     NeededSize = 0;
+  CONFIG_VAR_LIST_ENTRY      *Entry;
+
+  // Sanity check for input parameters
+  if ((VariableListBuffer == NULL) || (Size == NULL) || (VariableEntry == NULL)) {
+    Status = EFI_INVALID_PARAMETER;
+    goto Exit;
+  }
+
+  if (*Size < sizeof (*VarList)) {
+    Status = EFI_BUFFER_TOO_SMALL;
+    goto Exit;
+  }
+
+  // index into variable list
+  BinSize    = *Size;
+  VarList    = (CONST CONFIG_VAR_LIST_HDR *)((CHAR8 *)VariableListBuffer);
+  NeededSize = VAR_LIST_SIZE (VarList->NameSize, VarList->DataSize);
+
+  if (NeededSize > BinSize) {
+    // the NameSize and DataInBinSize have bad values and are pushing us past the end of the binary
+    DEBUG ((DEBUG_ERROR, "%a VarList buffer does not have needed size (actual: %x, expected: %x)\n", __FUNCTION__, BinSize, NeededSize));
+    *Size  = NeededSize;
+    Status = EFI_BUFFER_TOO_SMALL;
+    goto Exit;
+  }
+
+  // Use this as stub to indicate how much buffer used.
+  *Size = NeededSize;
+
+  /*
+    * Var List is in DmpStore format:
+    *
+    *  struct {
+    *    CONFIG_VAR_LIST_HDR VarList;
+    *    CHAR16 Name[VarList->NameSize/2];
+    *    EFI_GUID Guid;
+    *    UINT32 Attributes;
+    *    CHAR8 DataInBin[VarList-DataSize];
+    *    UINT32 CRC32; // CRC32 of all bytes from VarList to end of DataInBin
+    *  }
+    */
+  NameInBin  = (CONST CHAR16 *)(VarList + 1);
+  Guid       = (CONST EFI_GUID *)((CHAR8 *)NameInBin + VarList->NameSize);
+  Attributes = *(CONST UINT32 *)(Guid + 1);
+  DataInBin  = (CONST CHAR8 *)Guid + sizeof (*Guid) + sizeof (Attributes);
+  CRC32      = *(CONST UINT32 *)(DataInBin + VarList->DataSize);
+
+  // validate CRC32
+  CalcCRC32 = CalculateCrc32 ((VOID *)VarList, NeededSize - sizeof (CRC32));
+  if (CRC32 != CalcCRC32) {
+    DEBUG ((DEBUG_ERROR, "%a CRC is off in the variable list: actual: %x, expect %x\n", __FUNCTION__, CRC32, CalcCRC32));
+    Status = EFI_COMPROMISED_DATA;
+    goto Exit;
+  }
+
+  VarName = AllocatePool (VarList->NameSize);
+  if (VarName == NULL) {
+    DEBUG ((DEBUG_ERROR, "%a Failed to allocate memory for VarName size: %u\n", __FUNCTION__, VarList->NameSize));
+    Status = EFI_OUT_OF_RESOURCES;
+    goto Exit;
+  }
+
+  CopyMem (VarName, NameInBin, VarList->NameSize);
+
+  Data = AllocatePool (VarList->DataSize);
+  if (Data == NULL) {
+    DEBUG ((DEBUG_ERROR, "%a Failed to allocate memory for Data size: %u\n", __FUNCTION__, VarList->DataSize));
+    Status = EFI_OUT_OF_RESOURCES;
+    // Free VarName here, as other memory gets freed in exit routine, but we do not increment
+    // *ConfigVarListCount for this entry

```

Comment:
comment is out of date, referencing non-existent variables

Reply:
Updated. Thanks.